web_poet/exceptions/core.py

```diff
@@ -6,6 +6,19 @@
 """
 
 
+class InvalidInput(ValueError):
```

Comment:
How should we handle cases where a layout is invalid, and retry won't help, but fallback (e.g. to ML) can? Is it going to be a separate exception? What's the class hierarchy - are they separate, or is one of them a subclass of another?

Reply:
I was not planning to make it part of this pull request, but it does make sense to do so.

What about something in these lines? (I did not think too hard about names)

- `BadInput(ValueError)`
  - `WrongInputType`: the type of input data does not seem meant for this page object
    A fallback mechanism could make sense in a crawl targeting additional page objects, although from my experience ignoring such inputs is usually better (e.g. if a product URL points/redirects to a product list URL, you may not want to crawl the products on that product list).
  - `UnknownLayout`: the input seems meant for this page object, but extraction expectations (for key fields?) are failing, probably due to upstream layout changes or an A/B tests.
    A fallback mechanism to a machine-learning alternative would make sense.
  - `NotFound`: the resource is not present, e.g. HTTP 404 or HTTP 410.
    An HTTP 404 or HTTP 410 response may *not* raise this exception if it contains the expected data nonetheless, or if it contains data matching the other `BadInput` subclasses, or if there is reason to believe the response is a ban, meaning `Retry` should be raised instead.
    A fallback mechanism would not make sense.

Even if no fallback of any kind is implemented for some or any of these bad input exceptions, raising different exceptions may be helpful for stats.

In line with this, I wonder if we should implement a `RetriesExceeded` exception, to be thrown by web-poet frameworks after too many failed retries, and whether that exception should be a child of `BadInput` or not.

Reply:
See also: https://github.com/scrapinghub/web-poet/pull/26/files

I'm not sure how common is NotFound; what do you think about focussing on the first two now? 

Regarding names, there are at least two ways to approach it. First is to reflect action in the name: if fallback is needed, raise some kind of Fallback.. exception. Second is to be descriptive, and have the fallback logic implicit. 

It seems it'd be good to be explicit about fallbacks, to ensure users don't trigger it by mistake. For example, there might an item which extends standard Product item. If a layout is unknown, it could make sense to trigger a fallback, to fill at least the standard fields. In a second project there might be an item which is completely custom. Fallback doesn't make sense in this case, even if layout is unknown.

Reply:
What about `Abort(ValueError)` and `DelegateFallback(ValueError)`?

I wonder if a slightly shorter `UseFallback` or `FallBack` would work instead of `DelegateFallback`. Instead of `Abort` we could use `Fail` or `GiveUp` (I considered `End` and `Stop`, but besides verbs they are also names, so it seems less ideal).

I also wonder if `DelegateFallback` should be a subclass of `Abort`, to make it easier to handle both the same way on cases where a fallback is not supported.

Reply:
I'm okay with `UseFallback`. ~On the other hand, I think `AbortPageObject` would be much cleaner regarding what to abort.~ But rethinking about it, we're using `Retry` and not `RetryPageObject`.

In terms of exception hierarchies, should we have a base exception like the following to neatly tie them together?

```
Exception
├── RequestDownloaderVarError
└── ValueError
    └── PageObjectActions
        ├── Retry
        ├── UseFallback
        └── AbortPageObject
```

Reply:
Besides making it `PageObjectAction` (singular), :+1:

Reply:
Yeah, if the reason to have different exceptions is to handle them in different ways it makes more sense to have "fallback" and "abort" instead of explicit "the layout is unknown" (I assume there are more reasons to fallback than this one).

Reply:
Hey! One more thought, to complicate things, as usual :) Let's say we have a page object which extract product information, using zyte-common-items Product schema. We have the following error cases:

1. The page is not a product page, so Product can't be extracted.
2. The page is a product page, and page object doesn't know how to handle the layout. We expect page object to be able to handle all layouts on this website.
3. The page is a product page, and page object doesn't know how to handle the layout. We expect that there could be some layouts that the page object doesn't handle.
4. The page is blank or incorrect, we expect that a retry could help.

It seems this is what we should do in each of the cases:

1. We should return an empty Product, with metadata.probability set to 0. 
2. Not clear. Probably, we should raise a fallback error, and also send this page to developers for investigation.
3. Raise a fallback error.
4. Raise a Retry error.

It seems there is no scenario when page object should be raising an Abort exception, unless I'm missing something. Does it make sense? 

I'm not sure how to support the first case. Some options:

1. Page object should raise Abort exception. Returning an empty item with zero probability is handles elsewhere.
2. We allow to return an item from the validate_input method. If an item is returned from validate_input, then it's cached, and used in to_item and in all the fields.

Reply:
My thinking for `Abort` was that it would result in the errback being called.

I see how that would be problematic, though. For example, maybe you want to request both Product and ProductList, and even if one is wrong, you want your callback to receive both, and be able to tell which one was the right one, and for it to have the expected data.

However, I don’t see the point in returning an item from `validate_input`. How would that be different from not raising Abort and returning the item with probability=0 from `to_item`?

Right now I am thinking option 1 would be best, with a scrapy-poet implementation that calls the callback but gives `None` or some sentinel value for the requested item when an `Abort` exception is raised for it.

And maybe we can implement in zyte-common-items a spider middleware to transform that `None`/`sentinel` into an item with the response URL and 0 probability, although I am not sure in what scenarios that would be preferable.

Reply:
> However, I don’t see the point in returning an item from validate_input. How would that be different from not raising Abort and returning the item with probability=0 from to_item?

The difference is that in this case we can make fields work properly.

Reply:
Work properly meaning raising `Abort`?

Reply:
```py
@attrs.define
class MyPage(ItemPage[Product]):
    response: HttpResponse

    @field
    def price(self):
        return self.response.css("...")

    @field
    def metadata(self):
        return Metadata(url=self.response.url, probability=1.0)
    
    # ...
    
    def validate_input(self):
        if len(self.response.body) < 100:
            # looks like a download or rendering issue
            raise Retry()  

        if not self.response.css(".product").get():
            # not a product page, a right thing is to return probability=0 without trying to extract data
            return Product(url=self.response.url, metadata=Metadata(probability=0))
            
        if not self.respose.css(".product .supported-layout").get():
            # unsupported layout, use a fallback
            raise Fallback()

# later
page = create_page(MyPage, ...)
item = await page.to_item()

assert item.price == page.price
assert item.metadata.probability == page.metadata.probability
```

Reply:
Oh, I see, the fields of the output item of `validate_item` would get written into the corresponding fields, besides also becoming the output of `to_item`. (“used in to_item and in all the fields” as you described in option 2)

It definitely would not be a good idea to have the output of `to_item` be written back into fields, so I think the return-item approach may be the best idea.

I do wonder if it would be cleaner to go for a slightly different API: raising an exception with the item passed as a parameter, rather than returning the item (e.g. `raise Abort(item=…)`, or `raise SetItem(…)`). But I have no strong opinion on it, and they would work basically the same way in practice.

Reply:
Returning an item looks ok to me from the API point of view, because of these reasons:

1) it's similar to how scrapy's downloader middlewares work - they can either raise an exception, or return a new request or response
2) it's less code than having to wrap an item in an exception
3) you can have it type checked properly; I'm not sure how to express in mypy "this method should raise an Abort exception which argument must be of an ItemT type"